### ISSUE TITLE
Localize using existing map of environment

### ISSUE THREAD

**Issue by vkee** [penguinmenac3/ros_graph_slam#1]
Hi,

Is it possible to localize using an existing map generated by this system on a previous run through the same environment?

Thanks!

**Comment by penguinmenac3** [comment]
Short Answer: Theoretically yes, practically not implemented yet.

Long Answer:
Bad things first: To be honest I never thought about reloading a map, whereas the underlying graph slam architecture is not optimal for the kidnapped robot problem (relocalizing a robot on a map when you do not know where you are). To be even more specific I don't even know if there is a mathematically correct way of doing so already researched.

Maybe good thing: If your robot explores an environment creates a graph and you exactly know where you are when the robot is turned on again (e.g. where you turned it off) it is possible but not implemented yet. In this case you can implement a method that saves all edges and nodes of the graph (basically all operations done to it) to a log and then when starting (loading) the map, redo all the opperations from the log and add an odometry edge to the graph pointing to the location where the robot continues to map.

I guess this is not the answer you hoped to hear, but unfortunately this was never required by our research group. If you have any implementation or ideas for the kidnapped robot problem I would like to hear them and merge them!

**Comment by vkee** [comment]
I've been working on this (https://github.com/ethz-asl/segmatch/issues/47) and was looking to see if also you have been working on it to see if I could get some ideas. Maybe this might be useful for you?

**Comment by penguinmenac3** [comment]
I finally got the time to have a more detailed look into segmatch and it appears to be an awesome loop closure backend.

As I was thinking of your problem, I might found a quite simple solution to continue mapping with graph slam (at least in my implementation):
1. Map normally
2. When robot starts up again do not map until a loop closure was detected
3. Insert a loop closure edge from a node in the old graph to a newly created node
4. Initialize the new node with the estimate of the loop closure (initial estimate is important)
5. Continue mapping normally (the two graphs are tacked together)

=> Every further loop closure binds the two graphs stronger together and should (theoretically) improve the quallity.
The beauty of this approach is that any loop closure backend could be used to continue mapping of an existing map.

Thanks for making me think about that feature. I guess I will implement that over the next couple of weeks. Were you successfull on your implementation?

**Comment by vkee** [comment]
Yes, I was. What I did was I changed the prior when a localization was detected using the estimated transformation.

**Comment by penguinmenac3** [comment]
But how did you tie together the two graphs then? Do I assume correctly that you changed the prior of your new (relocalization) graph?

**Comment by vkee** [comment]
I didn't keep the old factor graph around. Yes.

You can see the link I posted to the issue where it describes in detail what I did.
